Camera3D Improvements

ffreyer opened the pull request:
# Description

Some general camera improvements/changes

#### Breaking

- revert the switch to fov based zoom. `Camera3D` now uses the same zoom style as `old_cam3d!` again, i.e. translating `eyeposition`. 
    - to keep clipping issues (#602, #941) from occurring  near and far are now relative to `norm(eyeposition - lookat)`, i.e. the zoom level. 
    - because of this stretch/contract view hotkeys have been removed as they are equivalent to zoom in/out. The old zoom in/out is now available as increase/decrease fov.
- Separate attributes into `cam.controls` for hotkeys and `cam.settings` for everything else

#### Non-breaking

- make near & far relative to `norm(eyeposition - lookat)`, i.e. zoom level
- scale translations with zoom level and fov. This normalizes the perceived translation speed 
- scale camera centering on bounding boxes with fov. This makes the full bbox visible at a larger range of fovs.
- add orbital `update_cam!(scene, cam, phi, theta[, radius, center])
- add deprecation warning for `old_cam3d!` etc. `Camera3D` can now reproduce this, though not exactly due to QoL tweaks.
- improve default hotkey choices
- add functionality to focus camera on a point on a plot

Fixes #2721, #1600, #1236 (y/z swap, documentation)

#### DataInspector fixes

I moved a bunch of code around that's also used by `DataInspector` so I did some testing and fixing on that too. Most of these fixes are probably not related to this pr.
- fixed error related to `getindex_sv(::Range, idx)`
- fixed blinking in `image` inspection
- fixed typos in arrows (closes #2764) & lines(egments) 
- fixed incorrect values when inspecting volumeslices
- fixed incorrect indicator placement in lines and added workaround to avoid multiple indicators

## Type of change

Breaking change

## Checklist

- [x] Added an entry in NEWS.md (for new features and breaking changes)
- [x] Added or changed relevant sections in the documentation
- [x] Check DataInspector 
- [x] Document changes to `ispressed`

ffreyer commented:
It would be good to get some feedback on how these changes feel or if there's anything else worth adding to make the 3D camera feel better.

For example I think that for orthographic cameras things are disappearing too quickly.

ffreyer commented:
Refimg tests all look fine. The differences should all just be minor changes in the zoom level from `center!()`

ffreyer commented:
At default fov the only major difference between `old_cam3d!` and `cam3d!` I can notice is that translations are too fast for the old one (around default zoom). Since the old cam doesn't scale translation speed with zoom this will get more different when changing fov. (Same with cad = true)

I think `Camera3D` is pretty much just a better, more feature rich version of `old_cam3d!` now.

Regarding tests - there are some small differences in the default zoom level I believe, which only GLMakie tests catch. So nothing to worry about.

asinghvi17 commented:
I just tried this on mac with touchpad, felt pretty good overall but some speeds definitely needed adjusting for trackpad.

Rotations were way to fast, and zoom was way too slow.  What worked for me was:
```julia
cam = cameracontrols(ax.scene)
cam.settings.mouse_rotationspeed[] = 0.5
cam.settings.mouse_zoomspeed[] = 2
```
which made things a lot more responsive.

ffreyer commented:
Touchpad Windows: I agree, your defaults feel better
Mouse (same) Windows: Rotation speed is ok either way I think, zoomspeed feels better at 1 for me.
Mouse Linux: I feel like `rotationspeed = 1` is already on the slow side and `zoomspeed  = 2` is on the faster side.

It seems like mouse speed is roughly 1:2 between windows and linux for me - matching that 0.5 factor. With a `scatter(rand(Point3f, 10))` I can do a 360° rotation by moving the cursor 1 window width with about 1.5 rotationspeed on linux and about 0.75 on windows. Not sure if that's a sane default to aim at, but maybe it helps normalizing things across OS's?

For zooming I think it's very dependent on trackpad vs mouse, which we can't really tell apart :slightly_frowning_face:

ffreyer commented:
I discussed this pr with @lazarusA at Makiecon. To summarize:
1. Some of the controls are inaccessible (home, page up, page down)
2. It would be nice to have a `focus on clicked position (on object)` like Blender does
3. It would be nice to have #2624 

In the last two commits I adressed point 1 and 2. The former fixes point 1 - camera reset now works via `left_control + left_click` and fov changes use b and n.

The latter commit added functionality for adjusting the point the camera focused via `left_alt + left_click`. This should work with any plot type except `text` and it should be an exact position for anything but scatter, meshscatter (these use plot.position) and volume (which uses the volume bounding box).

To get there I moved a bunch of code around so I'll need to do some more testing to check if everything works cleanly. I also extended `ispressed` so it can be used with `released` events. I still need to adjust the documentation for this.

For point 3 I want to open a separate pr.

ffreyer commented:
Alright I went through and tested DataInspector. I fixed a bunch of small issues which were probably unrelated.

The errors in GLMakie are related to minor changes in the default camera position/zoom. The GLMakie specific tests have higher precision and notice this, the normal refiimages do not. I adjusted the camera interaction tests in Makie with new numbers so that those don't fail.

I still need to make a few tweaks to documentation. After that this will be ready to merge again.

ffreyer commented on GLMakie/test/glmakie_refimages.jl:
```diff
@@ -65,6 +65,7 @@ end
         markersize=size,
         axis = (; scenekw = (;limits=Rect3f(Point3(0), Point3(1))))
     )
+    update_cam!(ax.scene, Point3f(2.224431, 2.224431, 2.128731), Point3f(0.5957, 0.5957, 0.50000006))
```
I hard coded the camera setup here (and below) to avoid test errors. I don't know if we want to keep this or update refimgs so I'm pointing it out. The difference is tiny.

ffreyer commented:
I'm closing this since it's part of the beta branch now
